{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "precise-transparency",
   "metadata": {
    "kernel": "Bash"
   },
   "source": [
    "# Bioinformatics workflow exercise: SoS and linear mixed model\n",
    "\n",
    "Author: Haoyue Shuai, Nov 17, 2020\n",
    "\n",
    "This tutorial introduces a workflow language, Script of Scripts (SoS), for bioinformatics analysis pipelines, with an example implementation of various linear mixed model methods for genetic association studies.\n",
    "\n",
    "This is an SoS Notebook with SoS kernel cells containing workflow steps written in SoS, and bash kernel cells to run these workflow steps. Please run bash codes here directly in this notebook so the output will be saved to the notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "impressive-summit",
   "metadata": {
    "kernel": "Bash"
   },
   "source": [
    "## Jupyter Lab setup\n",
    "\n",
    "Download this notebook and launch it with [JupyterLab](https://jupyter.org/). You can follow [these suggested setup instructions](http://statgen.us/lab-wiki/orientation/jupyter-setup.html).\n",
    "\n",
    "Please first making sure you have all the kernels needed. They should be available after all software are installed as instructed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "public-replica",
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Available kernels:\r\n",
      "  ir              /Users/hyeonjukim/Library/Jupyter/kernels/ir\r\n",
      "  julia-1.0       /Users/hyeonjukim/Library/Jupyter/kernels/julia-1.0\r\n",
      "  julia-1.5       /Users/hyeonjukim/Library/Jupyter/kernels/julia-1.5\r\n",
      "  calysto_bash    /Users/hyeonjukim/opt/miniconda3/share/jupyter/kernels/calysto_bash\r\n",
      "  python3         /Users/hyeonjukim/opt/miniconda3/share/jupyter/kernels/python3\r\n",
      "  sos             /Users/hyeonjukim/opt/miniconda3/share/jupyter/kernels/sos\r\n",
      "  python2         /usr/local/share/jupyter/kernels/python2\r\n",
      "\n"
     ]
    }
   ],
   "source": [
    "jupyter kernelspec list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "nominated-chemical",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "# parameter 1\n",
    "parameter: n = 1.0\n",
    "# parameter 2\n",
    "parameter: beta = [1.0,2.0,3.0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "average-oakland",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Print the value of n with bash\n",
    "[print_n]\n",
    "bash: expand = '${ }'\n",
    "    echo ${n}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "common-russia",
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Running \u001b[32mprint_n\u001b[0m: Print the value of n with bash\r\n",
      "1.0\r\n",
      "INFO: \u001b[32mprint_n\u001b[0m is \u001b[32mcompleted\u001b[0m.\r\n",
      "INFO: Workflow print_n (ID=w4bcbb8958466f710) is executed successfully with 1 completed step.\r\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sos run orientation-hkim.ipynb print_n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "normal-dynamics",
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Running \u001b[32mprint_n\u001b[0m: Print the value of n with bash\r\n",
      "666.0\r\n",
      "INFO: \u001b[32mprint_n\u001b[0m is \u001b[32mcompleted\u001b[0m.\r\n",
      "INFO: Workflow print_n (ID=we094e7d433abb2ad) is executed successfully with 1 completed step.\r\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sos run orientation-hkim.ipynb print_n --n 666"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "established-display",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[print_beta]\n",
    "bash: expand = '${ }'\n",
    "    echo ${beta}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "senior-python",
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Running \u001b[32mprint_beta\u001b[0m: \r\n",
      "[1.0, 2.0, 3.0]\r\n",
      "INFO: \u001b[32mprint_beta\u001b[0m is \u001b[32mcompleted\u001b[0m.\r\n",
      "INFO: Workflow print_beta (ID=w78fa93e094c77376) is executed successfully with 1 completed step.\r\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sos run orientation-hkim.ipynb print_beta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "fourth-ranch",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Print log(beta) with Python\n",
    "[log_beta]\n",
    "python: expand = '${ }'\n",
    "    import numpy as np\n",
    "    print(np.log(${beta}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "institutional-quarter",
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Running \u001b[32mlog_beta\u001b[0m: Print log(beta) with Python\r\n",
      "[0.         0.69314718 1.09861229]\r\n",
      "INFO: \u001b[32mlog_beta\u001b[0m is \u001b[32mcompleted\u001b[0m.\r\n",
      "INFO: Workflow log_beta (ID=w077f5b194f70ad1b) is executed successfully with 1 completed step.\r\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sos run orientation-hkim.ipynb log_beta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "embedded-conservative",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Print exp(n) with R\n",
    "[exp_n]\n",
    "R: expand = '${ }'\n",
    "    print(exp(${n}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "polyphonic-argentina",
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Running \u001b[32mexp_n\u001b[0m: Print exp(n) with R\r\n",
      "[1] 2.718282\r\n",
      "INFO: \u001b[32mexp_n\u001b[0m is \u001b[32mcompleted\u001b[0m.\r\n",
      "INFO: Workflow exp_n (ID=w68dec66676a24f9f) is executed successfully with 1 completed step.\r\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sos run orientation-hkim.ipynb exp_n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "convinced-brief",
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: sos run orientation-hkim.ipynb\r\n",
      "               [workflow_name | -t targets] [options] [workflow_options]\r\n",
      "  workflow_name:        Single or combined workflows defined in this script\r\n",
      "  targets:              One or more targets to generate\r\n",
      "  options:              Single-hyphen sos parameters (see \"sos run -h\" for details)\r\n",
      "  workflow_options:     Double-hyphen workflow-specific parameters\r\n",
      "\r\n",
      "Workflows:\r\n",
      "  print_n\r\n",
      "  print_beta\r\n",
      "  log_beta\r\n",
      "  exp_n\r\n",
      "\r\n",
      "Global Workflow Options:\r\n",
      "  --n 1.0 (as float)\r\n",
      "                        parameter 1\r\n",
      "  --beta 1.0 2.0 3.0 (as list)\r\n",
      "                        parameter 2\r\n",
      "\r\n",
      "Sections\r\n",
      "  print_n:              Print the value of n with bash\r\n",
      "  print_beta:\r\n",
      "  log_beta:             Print log(beta) with Python\r\n",
      "  exp_n:                Print exp(n) with R\r\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sos run orientation-hkim.ipynb -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "strong-allen",
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: sos run ../workflow/LMM.ipynb\r\n",
      "               [workflow_name | -t targets] [options] [workflow_options]\r\n",
      "  workflow_name:        Single or combined workflows defined in this script\r\n",
      "  targets:              One or more targets to generate\r\n",
      "  options:              Single-hyphen sos parameters (see \"sos run -h\" for details)\r\n",
      "  workflow_options:     Double-hyphen workflow-specific parameters\r\n",
      "\r\n",
      "Workflows:\r\n",
      "  boltlmm\r\n",
      "  gcta\r\n",
      "  fastGWA\r\n",
      "  regenie\r\n",
      "  SAIGE\r\n",
      "\r\n",
      "Global Workflow Options:\r\n",
      "  --cwd VAL (as path, required)\r\n",
      "                        the output directory for generated files\r\n",
      "  --sampleFile VAL (as path, required)\r\n",
      "                        Path to sample file\r\n",
      "  --bfile VAL (as path, required)\r\n",
      "                        Genotype files in plink binary this is used for\r\n",
      "                        computing the GRM\r\n",
      "  --genoFile  paths\r\n",
      "\r\n",
      "                        Path to bgen or bed files\r\n",
      "  --phenoFile VAL (as path, required)\r\n",
      "                        Phenotype file for quantitative trait (BMI)\r\n",
      "  --phenoCol VAL VAL ... (as type, required)\r\n",
      "                        Phenotype to be analyzed (specify the column)\r\n",
      "  --covarFile . (as path)\r\n",
      "                        Covariate file path. Will use phenoFile if empty\r\n",
      "  --formatFile . (as path)\r\n",
      "                        Summary statisticss format file path used for unifying\r\n",
      "                        output column names. Will not unify names if empty\r\n",
      "  --covarCol  (as list)\r\n",
      "                        Qualitative covariates to be used in the analysis\r\n",
      "  --qCovarCol  (as list)\r\n",
      "                        Quantitative covariates to be used in the analysis\r\n",
      "  --numThreads 2 (as int)\r\n",
      "                        Specific number of threads to use\r\n",
      "  --bgenMinMAF 0.001 (as float)\r\n",
      "                        Minimum MAF to be used\r\n",
      "  --bgenMinINFO 0.8 (as float)\r\n",
      "                        Mimimum info score to be used\r\n",
      "  --job-size 1 (as int)\r\n",
      "                        For cluster jobs, number commands to run per job\r\n",
      "  --container-lmm 'statisticalgenetics/lmm:1.6'\r\n",
      "                        The container with the lmm software. Can be either a\r\n",
      "                        dockerhub image or a singularity `sif` file. Default is\r\n",
      "                        set to using dockerhub image\r\n",
      "  --container-marp 'gaow/marp'\r\n",
      "\r\n",
      "Sections\r\n",
      "  boltlmm_1:            Run BOLT analysis\r\n",
      "    Workflow Options:\r\n",
      "      --covarMaxLevels VAL (as int, required)\r\n",
      "                        Maximum categories of covariates allowed\r\n",
      "      --LDscoresFile VAL (as path, required)\r\n",
      "                        Path to LDscore file for reference population\r\n",
      "      --geneticMapFile VAL (as path, required)\r\n",
      "                        Path to genetic map file used to interpolate genetic map\r\n",
      "                        coordinates from SNP physical (base pair) positions\r\n",
      "      --lmm-option lmm\r\n",
      "                        LMM option: lmm, lmmInfOnly, and lmmForceNonInf\r\n",
      "  gcta_1:               Partition the GRM into 100 parts and allocate 8GB memory\r\n",
      "                        to each job\r\n",
      "    Workflow Options:\r\n",
      "      --parts 100 (as int)\r\n",
      "                        Number of parts the GRM calculation is to be partitioned\r\n",
      "  gcta_2:               Merge all the parts together (Linux, Mac)\r\n",
      "  gcta_3:               Make a sparse GRM from the merged full-dense GRM\r\n",
      "  fastGWA_1:            fastGWA mixed model (based on the sparse GRM generated\r\n",
      "                        above)\r\n",
      "    Workflow Options:\r\n",
      "      --grmFile  path(f'{cwd}/{bfile:bn}.grm.sp')\r\n",
      "\r\n",
      "  regenie_0:            Select the SNPs and samples to be used based on maf,\r\n",
      "                        geno, hwe and mind options\r\n",
      "    Workflow Options:\r\n",
      "      --maf-filter 0.0 (as float)\r\n",
      "      --geno-filter 0.0 (as float)\r\n",
      "      --hwe-filter 0.0 (as float)\r\n",
      "      --mind-filter 0.0 (as float)\r\n",
      "  regenie_1:            Run REGENIE step 1: fitting the null\r\n",
      "    Workflow Options:\r\n",
      "      --bsize 400 (as int)\r\n",
      "                        Size of the genotype blocks to be used\r\n",
      "      --lowmem-prefix  cwd\r\n",
      "\r\n",
      "                        Path to temporarily store block predictions\r\n",
      "      --trait bt\r\n",
      "                        Specify that traits are binary with\r\n",
      "                        0=control,1=case,NA=missing (default is quantitative)\r\n",
      "  regenie_2:            Run REGENIE step 2: association analysis\r\n",
      "    Workflow Options:\r\n",
      "      --bsize 400 (as int)\r\n",
      "                        Size of the genotype blocks to be used\r\n",
      "      --minMAC VAL (as int, required)\r\n",
      "                        Mimimum allele count to be used\r\n",
      "      --trait bt\r\n",
      "  SAIGE_1:              Fit SAIGE null model\r\n",
      "    Workflow Options:\r\n",
      "      --trait-type VAL (as str, required)\r\n",
      "                        trait type, eg 'binary' or 'quantitative'\r\n",
      "      --loco TRUE\r\n",
      "                        Whether to use LOCO or not\r\n",
      "      --sampleCol IID\r\n",
      "                        Name of the sample column\r\n",
      "      --script-path /Users/hyeonjukim/software/bin/step1_fitNULLGLMM.R (as path)\r\n",
      "                        Path specific to SAIGE script\r\n",
      "      --invNormalize FALSE\r\n",
      "                        Inverse normalization only for non-normal quantitative\r\n",
      "                        traits\r\n",
      "  SAIGE_2:              Compute SAIGE statistics\r\n",
      "    Workflow Options:\r\n",
      "      --bgenMinMAC 4 (as int)\r\n",
      "                        Mimimum allele count to be used\r\n",
      "      --af-caco TRUE\r\n",
      "                        Specify whether to output allele frequencies in cases\r\n",
      "                        and controls\r\n",
      "      --script-path /Users/hyeonjukim/software/bin/step2_SPAtests.R (as path)\r\n",
      "                        Path specific to SAIGE script\r\n",
      "  boltlmm_2, fastGWA_2, SAIGE_3, regenie_3: Merge results and log files\r\n",
      "    Workflow Options:\r\n",
      "      --[no-]reverse-log-p (default to False)\r\n",
      "  boltlmm_3, fastGWA_3, SAIGE_4, regenie_4: Manhattan and QQ plots using `qqman`\r\n",
      "    Workflow Options:\r\n",
      "      --bp POS\r\n",
      "                        Column name for BP\r\n",
      "      --pval P\r\n",
      "                        Column name for p-value\r\n",
      "      --snp SNP\r\n",
      "                        Column name for SNP\r\n",
      "      --p-filter '0.05'\r\n",
      "                        Plot only on p-values smaller than this\r\n",
      "      --ylim 0 (as int)\r\n",
      "                        ylim set to 0 to use maximum -log10(p) in data\r\n",
      "  boltlmm_4, fastGWA_4, SAIGE_5, regenie_5: Generate analysis report: HTML file,\r\n",
      "                        and optionally PPTX file\r\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sos run ../workflow/LMM.ipynb -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "median-dictionary",
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LDSCORE.1000G_EUR.tab.gz\t\tgenotypes21_22.fam\r\n",
      "boltlmm_template.yml\t\t\timputed_genotypes.sample\r\n",
      "fastGWA_template.yml\t\t\timputed_genotypes_chr21.bgen\r\n",
      "genetic_map_hg19_withX.txt.gz\t\timputed_genotypes_chr21.bgen.bgi\r\n",
      "genotype_inventory.txt\t\t\timputed_genotypes_chr22.bgen\r\n",
      "genotypes.bed\t\t\t\timputed_genotypes_chr22.bgen.bgi\r\n",
      "genotypes.bim\t\t\t\tphenotypes.txt\r\n",
      "genotypes.fam\t\t\t\tregenie_template.yml\r\n",
      "genotypes21_22.bed\t\t\tregions.txt\r\n",
      "genotypes21_22.bim\t\t\tunrelated_samples.txt\r\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cd ../data/LMM_MWE\n",
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "sorted-wireless",
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "cd ~/GIT/orientation/notebook/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "italian-logistics",
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Running \u001b[32mfastGWA_1\u001b[0m: fastGWA mixed model (based on the sparse GRM generated above)\r\n",
      "INFO: Running \u001b[32mgcta_1\u001b[0m: Partition the GRM into 100 parts and allocate 8GB memory to each job\r\n",
      "HINT: Pulling docker image statisticalgenetics/lmm:1.6\r\n",
      "HINT: Docker image statisticalgenetics/lmm:1.6 is now up to date\r\n",
      "INFO: \u001b[32mgcta_1\u001b[0m (index=0) is \u001b[32mcompleted\u001b[0m.\r\n",
      "INFO: \u001b[32mgcta_1\u001b[0m (index=1) is \u001b[32mcompleted\u001b[0m.\r\n",
      "INFO: \u001b[32mgcta_1\u001b[0m output:   \u001b[32m../data/LMM_MWE/output/cache/genotypes.part_2_1.grm.bin ../data/LMM_MWE/output/cache/genotypes.part_2_1.grm.N.bin... (6 items in 2 groups)\u001b[0m\r\n",
      "INFO: Running \u001b[32mgcta_2\u001b[0m: Merge all the parts together (Linux, Mac)\r\n",
      "INFO: \u001b[32mgcta_2\u001b[0m is \u001b[32mcompleted\u001b[0m.\r\n",
      "INFO: \u001b[32mgcta_2\u001b[0m output:   \u001b[32m../data/LMM_MWE/output/genotypes.grm.bin ../data/LMM_MWE/output/genotypes.grm.N.bin... (3 items)\u001b[0m\r\n",
      "INFO: Running \u001b[32mgcta_3\u001b[0m: Make a sparse GRM from the merged full-dense GRM\r\n",
      "INFO: \u001b[32mgcta_3\u001b[0m is \u001b[32mcompleted\u001b[0m.\r\n",
      "INFO: \u001b[32mgcta_3\u001b[0m output:   \u001b[32m../data/LMM_MWE/output/genotypes.grm.sp\u001b[0m\r\n",
      "INFO: \u001b[32mfastGWA_1\u001b[0m (index=0) is \u001b[32mcompleted\u001b[0m.\r\n",
      "INFO: \u001b[32mfastGWA_1\u001b[0m (index=1) is \u001b[32mcompleted\u001b[0m.\r\n",
      "INFO: \u001b[32mfastGWA_1\u001b[0m output:   \u001b[32m../data/LMM_MWE/output/cache/imputed_genotypes_chr21.phenotypes.fastGWA.gz ../data/LMM_MWE/output/cache/imputed_genotypes_chr22.phenotypes.fastGWA.gz in 2 groups\u001b[0m\r\n",
      "INFO: Running \u001b[32mfastGWA_2\u001b[0m: Merge results and log files\r\n",
      "INFO: \u001b[32mfastGWA_2\u001b[0m is \u001b[32mcompleted\u001b[0m.\r\n",
      "INFO: \u001b[32mfastGWA_2\u001b[0m output:   \u001b[32m../data/LMM_MWE/output/phenotypes_BMI.fastGWA.snp_stats.gz ../data/LMM_MWE/output/phenotypes_BMI.fastGWA.snp_counts.txt\u001b[0m\r\n",
      "INFO: Running \u001b[32mfastGWA_3\u001b[0m: Manhattan and QQ plots using `qqman`\r\n",
      "INFO: \u001b[32mfastGWA_3\u001b[0m is \u001b[32mcompleted\u001b[0m.\r\n",
      "INFO: \u001b[32mfastGWA_3\u001b[0m output:   \u001b[32m../data/LMM_MWE/output/phenotypes_BMI.fastGWA.manhattan.png ../data/LMM_MWE/output/phenotypes_BMI.fastGWA.qq.png... (5 items)\u001b[0m\r\n",
      "INFO: Running \u001b[32mfastGWA_4\u001b[0m: Generate analysis report: HTML file, and optionally PPTX file\r\n",
      "HINT: Pulling docker image gaow/marp\r\n",
      "HINT: Docker image gaow/marp is now up to date\r\n",
      "INFO: \u001b[32mfastGWA_4\u001b[0m is \u001b[32mcompleted\u001b[0m.\r\n",
      "INFO: \u001b[32mfastGWA_4\u001b[0m output:   \u001b[32m../data/LMM_MWE/output/phenotypes_BMI.fastGWA.analysis_summary.html\u001b[0m\r\n",
      "INFO: Workflow fastGWA (ID=w005ee588ba40463d) is executed successfully with 7 completed steps and 9 completed substeps.\r\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sos run ../workflow/LMM.ipynb fastGWA \\\n",
    "    --cwd ../data/LMM_MWE/output \\\n",
    "    --bfile ../data/LMM_MWE/genotypes.bed \\\n",
    "    --sampleFile ../data/LMM_MWE/imputed_genotypes.sample \\\n",
    "    --genoFile ../data/LMM_MWE/imputed_genotypes_chr*.bgen \\\n",
    "    --phenoFile ../data/LMM_MWE/phenotypes.txt \\\n",
    "    --formatFile ../data/LMM_MWE/fastGWA_template.yml \\\n",
    "    --phenoCol BMI \\\n",
    "    --covarCol SEX \\\n",
    "    --qCovarCol AGE \\\n",
    "    --numThreads 1 \\\n",
    "    --bgenMinMAF 0.001 \\\n",
    "    --bgenMinINFO 0.1 \\\n",
    "    --parts 2 \\\n",
    "    --p-filter 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "generous-minimum",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "%preview ../data/LMM_MWE/output/phenotypes_BMI.fastGWA.manhattan.png"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "configured-young",
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[0;31mContinuation prompt found - input was incomplete:\n",
      "sos run ../workflow/LMM.ipynb regenie \\\n",
      "    --cwd ../data/LMM_MWE/output_regenie \\\n",
      "    --bfile ../data/LMM_MWE/genotypes21_22.bed \\\n",
      "    --maf-filter 0.001 \\\n",
      "    --sampleFile ../data/LMM_MWE/imputed_genotypes.sample \\\n",
      "    --genoFile ../data/LMM_MWE/imputed_genotypes_chr*.bgen \\\n",
      "    --phenoFile ../data/LMM_MWE/phenotypes.txt \\\n",
      "    --formatFile ../data/LMM_MWE/regenie_template.yml \\\n",
      "    --phenoCol ASTHMA T2D\\\n",
      "    --covarCol SEX \\\n",
      "    --qCovarCol AGE \\\n",
      "    --numThreads 4\\\n",
      "    --bsize 1000 \\\n",
      "    --lowmem_prefix ../data/LMM_MWE/output_regenie\\\n",
      "    --trait bt \\\n",
      "    --minMAC 4 \\\n",
      "    --bgenMinMAF 0.05 \\\n",
      "    --bgenMinINFO 0.8 \\\n",
      "    --reverse_log_p \\\n",
      "    --p-filter 1 \\\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "sos run ../workflow/LMM.ipynb regenie \\\n",
    "    --cwd ../data/LMM_MWE/output_regenie \\\n",
    "    --bfile ../data/LMM_MWE/genotypes21_22.bed \\\n",
    "    --maf-filter 0.001 \\\n",
    "    --sampleFile ../data/LMM_MWE/imputed_genotypes.sample \\\n",
    "    --genoFile ../data/LMM_MWE/imputed_genotypes_chr*.bgen \\\n",
    "    --phenoFile ../data/LMM_MWE/phenotypes.txt \\\n",
    "    --formatFile ../data/LMM_MWE/regenie_template.yml \\\n",
    "    --phenoCol ASTHMA T2D\\\n",
    "    --covarCol SEX \\\n",
    "    --qCovarCol AGE \\\n",
    "    --numThreads 4\\\n",
    "    --bsize 1000 \\\n",
    "    --lowmem_prefix ../data/LMM_MWE/output_regenie\\\n",
    "    --trait bt \\\n",
    "    --minMAC 4 \\\n",
    "    --bgenMinMAF 0.05 \\\n",
    "    --bgenMinINFO 0.8 \\\n",
    "    --reverse_log_p \\\n",
    "    --p-filter 1 \\\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "exotic-appliance",
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[0;31mContinuation prompt found - input was incomplete:\n",
      "sos run ../workflow/LMM1.ipynb regenie_2 \\\n",
      "    --cwd ../data/LMM_MWE/output_regenie \\\n",
      "    --bfile ../data/LMM_MWE/genotypes21_22.bed \\\n",
      "    --maf-filter 0.001 \\\n",
      "    --sampleFile ../data/LMM_MWE/imputed_genotypes.sample \\\n",
      "    --genoFile ../data/LMM_MWE/imputed_genotypes_chr*.bgen \\\n",
      "    --phenoFile ../data/LMM_MWE/phenotypes.txt \\\n",
      "    --formatFile ../data/LMM_MWE/regenie_template.yml \\\n",
      "    --phenoCol ASTHMA T2D\\\n",
      "    --covarCol SEX \\\n",
      "    --qCovarCol AGE \\\n",
      "    --numThreads 4\\\n",
      "    --bsize 1000 \\\n",
      "    --lowmem_prefix ../data/LMM_MWE/output_regenie\\\n",
      "    --trait bt \\\n",
      "    --minMAC 4 \\\n",
      "    --bgenMinMAF 0.05 \\\n",
      "    --bgenMinINFO 0.8 \\\n",
      "    --reverse_log_p \\\n",
      "    --p-filter 1 \\\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "sos run ../workflow/LMM1.ipynb regenie_2 \\\n",
    "    --cwd ../data/LMM_MWE/output_regenie \\\n",
    "    --bfile ../data/LMM_MWE/genotypes21_22.bed \\\n",
    "    --maf-filter 0.001 \\\n",
    "    --sampleFile ../data/LMM_MWE/imputed_genotypes.sample \\\n",
    "    --genoFile ../data/LMM_MWE/imputed_genotypes_chr*.bgen \\\n",
    "    --phenoFile ../data/LMM_MWE/phenotypes.txt \\\n",
    "    --formatFile ../data/LMM_MWE/regenie_template.yml \\\n",
    "    --phenoCol ASTHMA T2D\\\n",
    "    --covarCol SEX \\\n",
    "    --qCovarCol AGE \\\n",
    "    --numThreads 4\\\n",
    "    --bsize 1000 \\\n",
    "    --lowmem_prefix ../data/LMM_MWE/output_regenie\\\n",
    "    --trait bt \\\n",
    "    --minMAC 4 \\\n",
    "    --bgenMinMAF 0.05 \\\n",
    "    --bgenMinINFO 0.8 \\\n",
    "    --reverse_log_p \\\n",
    "    --p-filter 1 \\\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "metallic-chicago",
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[0;31mContinuation prompt found - input was incomplete:\n",
      "sos run ../workflow/LMM.ipynb boltlmm \\\n",
      "    --cwd ./output_boltlmm \\\n",
      "    --bfile ../data/LMM_MWE/genotypes.bed \\\n",
      "    --sampleFile ../data/LMM_MWE/imputed_genotypes.sample \\\n",
      "    --genoFile ../data/LMM_MWE/imputed_genotypes_chr*.bgen \\\n",
      "    --phenoFile ../data/LMM_MWE/phenotypes.txt \\\n",
      "    --formatFile ../data/LMM_MWE/boltlmm_template.yml \\\n",
      "    --LDscoresFile ../data/LMM_MWE/LDSCORE.1000G_EUR.tab.gz \\\n",
      "    --geneticMapFile ../data/LMM_MWE/genetic_map_hg19_withX.txt.gz \\\n",
      "    --phenoCol BMI \\\n",
      "    --covarCol SEX \\\n",
      "    --covarMaxLevels 10 \\\n",
      "    --qCovarCol AGE \\\n",
      "    --numThreads 5 \\\n",
      "    --bgenMinMAF 0.001 \\\n",
      "    --bgenMinINFO 0.1 \\\n",
      "    --lmm-option none \\\n",
      "    --p-filter 1 \\\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "sos run ../workflow/LMM1.ipynb boltlmm \\\n",
    "    --cwd ./output_boltlmm \\\n",
    "    --bfile ../data/LMM_MWE/genotypes.bed \\\n",
    "    --sampleFile ../data/LMM_MWE/imputed_genotypes.sample \\\n",
    "    --genoFile ../data/LMM_MWE/imputed_genotypes_chr*.bgen \\\n",
    "    --phenoFile ../data/LMM_MWE/phenotypes.txt \\\n",
    "    --formatFile ../data/LMM_MWE/boltlmm_template.yml \\\n",
    "    --LDscoresFile ../data/LMM_MWE/LDSCORE.1000G_EUR.tab.gz \\\n",
    "    --geneticMapFile ../data/LMM_MWE/genetic_map_hg19_withX.txt.gz \\\n",
    "    --phenoCol BMI \\\n",
    "    --covarCol SEX \\\n",
    "    --covarMaxLevels 10 \\\n",
    "    --qCovarCol AGE \\\n",
    "    --numThreads 5 \\\n",
    "    --bgenMinMAF 0.001 \\\n",
    "    --bgenMinINFO 0.1 \\\n",
    "    --lmm-option none \\\n",
    "    --p-filter 1 \\\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "seven-wrapping",
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/hyeonjukim/GIT/orientation/notebook\r\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pwd"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "Bash",
     "calysto_bash",
     "Bash",
     "#E6EEFF",
     ""
    ],
    [
     "SoS",
     "sos",
     "",
     "",
     "sos"
    ]
   ],
   "panel": {
    "displayed": true,
    "height": 0
   },
   "version": "0.20.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
